File: 1 - LinkedList/Java/3-CircularSingleLinkedList/circularSingleLinkedList.java
public class circularSingleLinkedList {
    // Node class representing each element in the circular singly linked list
    static class Node {
        String id; // Student ID
        String name; // Student name
        String major; // Student major
        Node next; // Pointer to the next node in the list

        // Constructor to initialize a new node
        Node(String id, String name, String major) {
            this.id = id;
            this.name = name;
            this.major = major;
            this.next = null;
        }
    }

    // LinkedList class to manage the circular singly linked list
    static class LinkedList {
        private Node head; // Pointer to the head of the list
        private Node tail; // Pointer to the tail of the list
        private int size; // Size of the list

        // Constructor to initialize an empty list
        LinkedList() {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }

        // Function to create a circular singly linked list with the first node
        void createCircularSL(String id, String name, String major) {
            Node newNode = new Node(id, name, major);
            newNode.next = newNode;
            head = newNode;
            tail = newNode;
            size = 1;
        }

        // Function to add a node at the beginning of the list
        void addAtFirstNode(String id, String name, String major) {
            if (size == 0) {
                createCircularSL(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            newNode.next = head;
            tail.next = newNode;
            head = newNode;
            size++;
        }

        // Function to add a node at the nth position in the list
        void addAtNthNode(String id, String name, String major, int pos) {
            if (pos < 1 || pos > size + 1) {
                System.out.println("Invalid position!");
                return;
            }
            if (pos == 1) {
                addAtFirstNode(id, name, major);
                return;
            }
            if (pos == size + 1) {
                addAtLastNode(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            Node cur = head;
            for (int i = 1; i < pos - 1; i++) {
                cur = cur.next;
            }
            newNode.next = cur.next;
            cur.next = newNode;
            size++;
        }

        // Function to add a node at the end of the list
        void addAtLastNode(String id, String name, String major) {
            if (size == 0) {
                createCircularSL(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            newNode.next = head;
            tail.next = newNode;
            tail = newNode;
            size++;
        }

        // Function to update the first node in the list
        void updateFirstNode(String id, String name, String major) {
            if (size == 0) {
                System.out.println("List is empty!");
                return;
            }
            head.id = id;
            head.name = name;
            head.major = major;
        }

        // Function to update the nth node in the list
        void updateNthNode(String id, String name, String major, int pos) {
            if (pos < 1 || pos > size) {
                System.out.println("Invalid position!");
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos; i++) {
                cur = cur.next;
            }
            cur.id = id;
            cur.name = name;
            cur.major = major;
        }

        // Function to update the last node in the list
        void updateLastNode(String id, String name, String major) {
            if (size == 0) {
                System.out.println("List is empty!");
                return;
            }
            tail.id = id;
            tail.name = name;
            tail.major = major;
        }

        // Function to remove the first node from the list
        void removeFirstNode() {
            if (size == 0) {
                System.out.println("List is empty!");
                return;
            }
            if (size == 1) {
                head = null;
                tail = null;
            } else {
                head = head.next;
                tail.next = head;
            }
            size--;
        }

        // Function to remove the nth node from the list
        void removeNthNode(int pos) {
            if (pos < 1 || pos > size) {
                System.out.println("Invalid position!");
                return;
            }
            if (pos == 1) {
                removeFirstNode();
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos - 1; i++) {
                cur = cur.next;
            }
            cur.next = cur.next.next;
            if (pos == size) {
                tail = cur;
            }
            size--;
        }

        // Function to remove the last node from the list
        void removeLastNode() {
            if (size == 0) {
                System.out.println("List is empty!");
                return;
            }
            if (size == 1) {
                head = null;
                tail = null;
            } else {
                Node cur = head;
                while (cur.next != tail) {
                    cur = cur.next;
                }
                tail = cur;
                tail.next = head;
            }
            size--;
        }

        // Function to print the circular singly linked list
        void printCircularSL() {
            if (size == 0) {
                System.out.println("List is empty!");
                return;
            }
            System.out.println("Number of Nodes: " + countNodes()); // Display the total number of nodes
            Node cur = head;
            int i = 1;
            do {
                System.out.println("Node " + i);
                System.out.println("ID\t\t: " + cur.id);
                System.out.println("Name\t\t: " + cur.name);
                System.out.println("Major\t\t: " + cur.major + "\n");
                cur = cur.next;
                i++;
            } while (cur != head);
        }

        // Utility function to count the number of nodes in the list
        int countNodes() {
            return size; // Return the total number of nodes
        }
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        // Create the circular singly linked list with the first node
        list.createCircularSL("11223399", "Arief F-sa", "Computer Science");
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Add a node at the beginning of the list
        list.addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Add a node at the end of the list
        list.addAtLastNode("11223400", "Pejeel", "Drop out");
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Remove the first node from the list
        list.removeFirstNode();
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Remove the last node from the list
        list.removeLastNode();
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Add nodes at the beginning and end of the list
        list.addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
        list.addAtLastNode("11223400", "Pejeel", "Drop out");
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Add nodes at specific positions in the list
        list.addAtNthNode("11223401", "dummy1", "Math", 3);
        list.addAtNthNode("11223402", "dummy2", "Physics", 2);
        list.printCircularSL();
        System.out.println("==========\n\n\n");

        // Remove the node at the specified position in the list
        list.removeNthNode(2);
        list.printCircularSL();
        System.out.println("==========\n\n\n");
    }
}

File: 1 - LinkedList/C#/3-CircularSingleLinkedList/circularSingleLinkedList.cs
using System;

namespace CircularSingleLinkedList
{
    // Node class representing each element in the circular singly linked list
    public class Node
    {
        public string Id { get; set; }    // Student ID
        public string Name { get; set; }  // Student name
        public string Major { get; set; } // Student major
        public Node Next { get; set; }    // Pointer to the next node in the list

        // Constructor to initialize a new node
        public Node(string id, string name, string major)
        {
            Id = id;
            Name = name;
            Major = major;
            Next = null;
        }
    }

    // LinkedList class to manage the circular singly linked list
    public class LinkedList
    {
        private Node head; // Pointer to the head of the list
        private Node tail; // Pointer to the tail of the list
        private int size;  // Size of the list

        // Constructor to initialize an empty list
        public LinkedList()
        {
            head = null;
            tail = null;
            size = 0;
        }

        // Function to create a circular singly linked list with the first node
        public void CreateCircularSL(string id, string name, string major)
        {
            Node newNode = new Node(id, name, major);
            newNode.Next = newNode;
            head = newNode;
            tail = newNode;
            size = 1;
        }

        // Function to add a node at the beginning of the list
        public void AddAtFirstNode(string id, string name, string major)
        {
            if (size == 0)
            {
                CreateCircularSL(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            newNode.Next = head;
            tail.Next = newNode;
            head = newNode;
            size++;
        }

        // Function to add a node at the nth position in the list
        public void AddAtNthNode(string id, string name, string major, int pos)
        {
            if (pos < 1 || pos > size + 1)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            if (pos == 1)
            {
                AddAtFirstNode(id, name, major);
                return;
            }
            if (pos == size + 1)
            {
                AddAtLastNode(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            Node cur = head;
            for (int i = 1; i < pos - 1; i++)
            {
                cur = cur.Next;
            }
            newNode.Next = cur.Next;
            cur.Next = newNode;
            size++;
        }

        // Function to add a node at the end of the list
        public void AddAtLastNode(string id, string name, string major)
        {
            if (size == 0)
            {
                CreateCircularSL(id, name, major);
                return;
            }
            Node newNode = new Node(id, name, major);
            newNode.Next = head;
            tail.Next = newNode;
            tail = newNode;
            size++;
        }

        // Function to update the first node in the list
        public void UpdateFirstNode(string id, string name, string major)
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            head.Id = id;
            head.Name = name;
            head.Major = major;
        }

        // Function to update the nth node in the list
        public void UpdateNthNode(string id, string name, string major, int pos)
        {
            if (pos < 1 || pos > size)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos; i++)
            {
                cur = cur.Next;
            }
            cur.Id = id;
            cur.Name = name;
            cur.Major = major;
        }

        // Function to update the last node in the list
        public void UpdateLastNode(string id, string name, string major)
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            tail.Id = id;
            tail.Name = name;
            tail.Major = major;
        }

        // Function to remove the first node from the list
        public void RemoveFirstNode()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Node del = head;
            if (size == 1)
            {
                head = null;
                tail = null;
            }
            else
            {
                head = head.Next;
                tail.Next = head;
            }
            size--;
        }

        // Function to remove the nth node from the list
        public void RemoveNthNode(int pos)
        {
            if (pos < 1 || pos > size)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            if (pos == 1)
            {
                RemoveFirstNode();
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos - 1; i++)
            {
                cur = cur.Next;
            }
            Node del = cur.Next;
            cur.Next = del.Next;
            if (del == tail)
            {
                tail = cur;
            }
            size--;
        }

        // Function to remove the last node from the list
        public void RemoveLastNode()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Node del = tail;
            if (size == 1)
            {
                head = null;
                tail = null;
            }
            else
            {
                Node cur = head;
                while (cur.Next != tail)
                {
                    cur = cur.Next;
                }
                tail = cur;
                tail.Next = head;
            }
            size--;
        }

        // Function to print the circular singly linked list
        public void PrintCircularSL()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Console.WriteLine("Number of Nodes: " + size); // Display the total number of nodes

            Node cur = head;
            int i = 1;
            do
            {
                Console.WriteLine("Node " + i);
                Console.WriteLine("ID\t\t: " + cur.Id);
                Console.WriteLine("Name\t\t: " + cur.Name);
                Console.WriteLine("Major\t\t: " + cur.Major);
                Console.WriteLine();
                cur = cur.Next;
                i++;
            } while (cur != head);
        }

        // Utility function to count the number of nodes in the list
        public int CountNodes()
        {
            return size; // Return the total number of nodes
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            LinkedList list = new LinkedList();

            // Create the circular singly linked list with the first node
            list.CreateCircularSL("11223399", "Arief F-sa", "Computer Science");
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Add a node at the beginning of the list
            list.AddAtFirstNode("11223398", "Nisa Uswa", "Psychology");
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Add a node at the end of the list
            list.AddAtLastNode("11223400", "Pejeel", "Drop out");
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Remove the first node from the list
            list.RemoveFirstNode();
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Remove the last node from the list
            list.RemoveLastNode();
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Add nodes at the beginning and end of the list
            list.AddAtFirstNode("11223398", "Nisa Uswa", "Psychology");
            list.AddAtLastNode("11223400", "Pejeel", "Drop out");
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Add nodes at specific positions in the list
            list.AddAtNthNode("11223401", "dummy1", "Math", 3);
            list.AddAtNthNode("11223402", "dummy2", "Physics", 2);
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");

            // Remove the node at the specified position in the list
            list.RemoveNthNode(2);
            list.PrintCircularSL();
            Console.WriteLine("==========\n\n\n");
        }
    }
}